Add DO option to clipboard export, refs #13395

PeteVox opened the pull request:
Add a new system setting to enable export of digital objects attached to
the clipboard (information object and actor records only). All digital
object types can be exported, but only masters are exported for any
image type.

Digital objects are included in the export ZIP along with associated
metadata.

Incorporate basic user permissions checks prior to export, with further
checks for rights as part of the background job/task. Add a new method
to the digital object model to check for any conditional copyright
attached. If set, then do not export the affected digital object.

Modify the export page interface to rebuild options panel depending on
controller-based permissions checks. Also move help and other message
construction from template to controller.

Make the selection of descendant or digital objects in the export page
mutually exclusive. Add several new styling rules (LESS) to the Dominium
theme in scaffolding that apply to the options panel (#export-options).

Change some help and alert messages to improve consistency.

Build additional export form parameter validation in the controller to
mitigate attempts to manually override parameters, and also to prevent
direct access to the export page URL without objects on the clipboard.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -24,19 +24,47 @@ class ObjectExportAction extends DefaultEditAction
     $this->objectType = $request->getParameter('objectType');
+    
+    $this->descendantsIncluded = ('on' == $request->getParameter('includeDescendants') ? true : false );
+    $this->descendantsAllLevels = ($this->descendantsIncluded && 'on' == $request->getParameter('includeAllLevels') ? true : false );
```
Cosmetic: you could assign the comparison directly ...

```php
$this->descendantsIncluded = ('on' == $request->getParameter('includeDescendants'));
$this->descendantsAllLevels = ($this->descendantsIncluded && 'on' == $request->getParameter('includeAllLevels'));
```

The parenthesis may not be needed (specially in the first case), but I'm not completely sure.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -24,19 +24,47 @@ class ObjectExportAction extends DefaultEditAction
+    else
+    {
+      $this->draftsIncluded = false;
+    }
```
Cosmetic:

```php
$this->draftsIncluded = false;

if ($this->context->user->isAuthenticated() && 'on' == $request->getParameter('includeDrafts'))
{
  $this->draftsIncluded = true;
}
```

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -24,19 +24,47 @@ class ObjectExportAction extends DefaultEditAction
+    else
+    {
+      $this->includeDigitalObjects = false;
+    }
```
Cosmetic: similar to the comment above for the `$this->includeDigitalObjects` assignation.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -24,19 +24,47 @@ class ObjectExportAction extends DefaultEditAction
+      $this->includeDigitalObjects = false;
+    }
+    
+    $this->showOptions = ( 'actor' == $this->objectType && !$this->digitalObjectsAvailable ? false : true );
```
I think you need to consider in here that `$this->objectType` could also be `'repository'`.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -116,8 +126,64 @@ protected function addField($name)
+
+      case 'includeDescendants':
+          $this->form->setWidget('includeDescendants', new sfWidgetFormInputCheckbox(array('label' => 'Include descendants', 'help' => __('Choosing "Include descendants" will include all lower-level records beneath those currently on the clipboard in the export.'))));
+        break;
```
Cosmetic:

- Fix indentation, reduce to 2 spaces on the first line.
- Split array items in different lines to reduce line length.
- Empty line before `break;`.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -116,8 +126,64 @@ protected function addField($name)
+
+      case 'includeAllLevels':
+          $this->form->setWidget('includeAllLevels', new sfWidgetFormInputCheckbox(array('label' => 'Include all descendant levels of description')));
+        break;
```
Cosmetic: same as above.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -116,8 +126,64 @@ protected function addField($name)
+        break;
+
+      case 'includeDescendants':
+          $this->form->setWidget('includeDescendants', new sfWidgetFormInputCheckbox(array('label' => 'Include descendants', 'help' => __('Choosing "Include descendants" will include all lower-level records beneath those currently on the clipboard in the export.'))));
```
The label is missing the i18n helper.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -116,8 +126,64 @@ protected function addField($name)
+        break;
+
+      case 'includeAllLevels':
+          $this->form->setWidget('includeAllLevels', new sfWidgetFormInputCheckbox(array('label' => 'Include all descendant levels of description')));
```
The label is missing the i18n helper.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -116,8 +126,64 @@ protected function addField($name)
+          $size = 4;
+        }
+
+        $this->form->setWidget('levels', new sfWidgetFormSelect(array('label' => __('Select levels of descendant descriptions for inclusion'), 'help' => __('If no levels are selected, the export will fail. You can use the control (Mac ⌘) and/or shift keys to multi-select values from the Levels of description menu. It is necessary to include the level(s) above the desired export level, up to and i ...
```
Cosmetic: it would be nice to reduce this line length concatenating the text with `.` splitting the text in different lines.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -116,8 +126,64 @@ protected function addField($name)
+      case 'includeDigitalObjects':
+        if ($this->digitalObjectsAvailable)
+        {
+          switch ($this->objectType)
+          {
+            case 'actor':
+              $helpMessage = '';
+              break;
+
+            default:
+              $helpMessage = __('Please note: digital objects will not be exported for any included descendant records, only for records that are on the clipboard.');
+          }
```
Not considering 'repository' in this `switch`, which is probably cleaner as an `if`.

PeteVox replied:
Have a couple of switches in the revised code (to be added to repo) that validate and set some default parameters. After you've been able to take a look at the revised code, no problem at all to convert to if/then if you prefer.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -116,8 +126,64 @@ protected function addField($name)
+          }
+          
+          $this->form->setWidget('includeDigitalObjects', new sfWidgetFormInputCheckbox(array('label' => 'Include digital objects', 'help' => $helpMessage)));
+        } else {
```
Cosmetic: new lines for the `else` and the `{`.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -116,8 +126,64 @@ protected function addField($name)
+          
+          $this->form->setWidget('includeDigitalObjects', new sfWidgetFormInputCheckbox(array('label' => 'Include digital objects', 'help' => $helpMessage)));
+        } else {
+          $this->form->setWidget('includeDigitalObjects', new sfWidgetFormInputHidden(array('label' => false)));
```
There is a JS file that takes care of hiding/showing fields based on another input values. I think we need to reconsider that JS (I'll follow-up about this later).

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -116,8 +126,64 @@ protected function addField($name)
+          $this->form->setWidget('includeDigitalObjects', new sfWidgetFormInputHidden(array('label' => false)));
+        }
+        $this->form->setDefault('includeDigitalObjects', false);
+        break;
```
Cosmetic: empty line before `break`.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -116,8 +126,64 @@ protected function addField($name)
+              $helpMessage = __('Please note: digital objects will not be exported for any included descendant records, only for records that are on the clipboard.');
+          }
+          
+          $this->form->setWidget('includeDigitalObjects', new sfWidgetFormInputCheckbox(array('label' => 'Include digital objects', 'help' => $helpMessage)));
```
The label is missing the i18n helper.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -116,8 +126,64 @@ protected function addField($name)
+      case 'includeDrafts':
+        if ($this->context->user->isAuthenticated())
+        {
+          $this->form->setWidget('includeDrafts', new sfWidgetFormInputCheckbox(array('label' => 'Include draft records', 'help' => __('Choosing "Include draft records" will include those marked with a Draft publication status in the export. Note: if you do NOT choose this option, any descendants of a draft record will also be excluded, even if they are published.'))));
```
The label is missing the i18n helper.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -116,8 +126,64 @@ protected function addField($name)
+        } else {
+          $this->form->setWidget('includeDrafts', new sfWidgetFormInputHidden(array('label' => false)));
+        }
+        $this->form->setDefault('includeDrafts', false);
```
Cosmetic: similar issues than above with `else` and `break`.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -161,16 +227,17 @@ protected function doBackgroundExport($request)
-      'current-level-only' => 'on' !== $request->getParameter('includeDescendants'),
-      'public' => 'on' !== $request->getParameter('includeDrafts'),
+      'current-level-only' => !$this->descendantsIncluded,
+      'public' => $this->draftsIncluded,
```
Should this be `!$this->draftsIncluded`?

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -161,16 +227,17 @@ protected function doBackgroundExport($request)
-    if ('xml' === $this->format)
-    {
-      $options['name'] = $this->context->i18n->__('XML export');
+    $options['name'] = $this->context->i18n->__( ('xml' === $this->format ? 'XML export' : 'CSV export' ) );
```
Cosmetic: no need for spaces between parenthesis.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -161,16 +227,17 @@ protected function doBackgroundExport($request)
-      $options['name'] = $this->context->i18n->__('XML export');
+    $options['name'] = $this->context->i18n->__( ('xml' === $this->format ? 'XML export' : 'CSV export' ) );
+
+    if($this->includeDigitalObjects){
```
Cosmetic: space after `if` and new line for the `{`.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -161,16 +227,17 @@ protected function doBackgroundExport($request)
+    $options['name'] = $this->context->i18n->__( ('xml' === $this->format ? 'XML export' : 'CSV export' ) );
+
+    if($this->includeDigitalObjects){
+      $options['name'] = $this->context->i18n->__('%1% and', array('%1%' => sfConfig::get('app_ui_label_digitalobject'))) . ' ' . $options['name'];
```
`'%1% and %2%'` and including the name in the array will probably be clearer for translators.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -182,16 +249,18 @@ protected function doBackgroundExport($request)
       else
       {
-        $job = QubitJob::runJob($this->getJobNameString(), $options);
+        $job = QubitJob::runJob($jobName, $options);
```
Nice!

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -254,4 +323,125 @@ private function getJobNameString()
+   *
+   * @return boolean:digitalObjectsAvailable
+   */
+  protected function getDigitalObjects()
```
Cosmetic: with the current implementation, I think it makes more sense to call it `loadDigitalObjects()`, do not return and check the global vars outside after it's called.

PeteVox replied:
This now moved to _arBaseJob.class.php_ as part of background processing and does now return the valid digital object set. Will keep existing for the moment if ok - but no problem to change after you've been able to review the updated version.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -254,4 +323,125 @@ private function getJobNameString()
+    $this->digitalObjects = array();
+    
+    // process if export option is set and this is a description or actor export
+    if (sfConfig::get('app_clipboard_export_digitalobjects_enabled') && ('informationObject' == $this->objectType || 'actor' == $this->objectType))
```
It's better to use a default when possible (just in case):

```php
sfConfig::get('app_clipboard_export_digitalobjects_enabled', false);
```

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -254,4 +323,125 @@ private function getJobNameString()
+   */
+  protected function getDigitalObjects()
+  {
+    // set default value
```
Cosmetic: uppercase first letter in new line comments (we're very inconsistent with this).

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -254,4 +323,125 @@ private function getJobNameString()
+      switch ($this->objectType)
+      {
+        case 'informationObject':
+          $criteria->addJoin(QubitInformationObject::ID, QubitDigitalObject::OBJECT_ID);
```
This join is unnecessary.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -254,4 +323,125 @@ private function getJobNameString()
+          break;
+
+        case 'actor':
+          $criteria->addJoin(QubitActor::ID, QubitDigitalObject::OBJECT_ID);
```
This join is unnecessary.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -254,4 +323,125 @@ private function getJobNameString()
+          // ----------
+          // Do appropriate ACL check(s). Master copy of text objects are always allowed for reading
+          // QubitActor does not have a ACL check for readmaster - so only enable for authenticated users.
+          if ( QubitTerm::TEXT_ID == $obj->mediaTypeId 
```
Shouldn't it be `$digitalObject->mediaTypeId `?

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -254,4 +323,125 @@ private function getJobNameString()
+          // QubitActor does not have a ACL check for readmaster - so only enable for authenticated users.
+          if ( QubitTerm::TEXT_ID == $obj->mediaTypeId 
+              || (
+                $digitalObject->masterAccessibleViaUrl() 
```
Should this check affect text DO and information obj DO too?

PeteVox replied:
Absolutely. Fixed.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -254,4 +323,125 @@ private function getJobNameString()
+              || (
+                $digitalObject->masterAccessibleViaUrl() 
+                && ('actor' == $this->objectType && $this->context->user->isAuthenticated() && QubitAcl::check($item, 'read')) 
+                || ('informationObject' == $this->objectType && (QubitAcl::check($item, 'readMaster') && QubitGrantedRight::checkPremis($item->id, 'readMaster') && !$this->isCopyrightedMaster($item->id)))
```
Cosmetic: the way the parenthesis and indentation are used in here confuses a bit, making you believe the initial && and || above belong to the same group when the don't.

PeteVox replied:
Have simplified the logic in latest code. Still not sure the layout of parenthesis is ideal, but can edit after you've been able to see it.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -254,4 +323,125 @@ private function getJobNameString()
+        }
+      }
+    }
+    if (!empty($this->digitalObjects))
```
Cosmetic: empty line before control statement.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -254,4 +323,125 @@ private function getJobNameString()
+    {
+      $this->digitalObjectsAvailable = true;
+    }
+    return $this->digitalObjectsAvailable;
```
Cosmetic: if you keep the `return`, empty line before it.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -254,4 +323,125 @@ private function getJobNameString()
+   *
+   * @return boolean
+   */
+  protected function isCopyrightedMaster($objId)
```
I think this is a nice to have in the `QubitDigitalObject` model class as a public static function. And maybe reuse it from /apps/modules/digitalobject/actions/viewAction.class.php too.

PeteVox replied:
Done.

PeteVox replied:
Done.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -254,4 +323,125 @@ private function getJobNameString()
+    {
+      throw new sfException('Unexpected error');
+    }
+    return ( $r->has ? true : false );
```
Cosmetic: empty line before return and you could return `$r->has` directly.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -254,4 +323,125 @@ private function getJobNameString()
+
+    if (false === $r || !isset($r->has))
+    {
+      throw new sfException('Unexpected error');
```
Apart from a connection error or a malformed database, what could cause this? Not sure if throwing an exception is a good idea in here, maybe just return `true`?

PeteVox replied:
Not sure about the need for this check at all - but copied from original version in `QubitDigitalObject`. Have updated in revised code.

jraddaoui commented on apps/qubit/modules/object/templates/exportSuccess.php:
```diff
@@ -17,63 +17,90 @@
 
     <?php echo $form->renderHiddenFields() ?>
 
+    <style>
+      .export-options-row {
+        overflow: hidden;
+      }
+      .export-options-row-l {
+        width: 35%;
+        float: left;
+      }
+      .export-options-row-r {
+        width: 65%;
+        float: right;
+        position: relative;
+      }
+      .export-options-row .alert {
+        margin: 0 0 15px 0;
+        padding: 3px 6px;
+      }
+      .export-options-row-l, 
+      .export-options-row-l > *, 
+      .export-options-row-r {
+        display: inline-block;
+      }
+      #exportAllLevels {
+        padding-bottom: 10px;
+      }
+      #exportLevels {
+        margin-bottom: 15px;
+      }
+      #exportLevels label {
+        font-weight: bold;
+        padding-bottom: 5px;
+      }
+      #exportLevels .export-options-row-l {
+        width: 33%;
+      }
+    </style>
```
It would be nice to move this to https://github.com/artefactual/atom/blob/qa/2.6.x/plugins/arDominionPlugin/css/less/scaffolding.less.

And, if possible, use a bit of LESS structure, but it's not a problem if it's added as plain CSS. Ideally, we'd use as much of what's available in Bootstrap to reduce it and I think you could achieve most of what is done with the `row` and `span*` classes. Not a huge deal either.

PeteVox replied:
This has been changed completely in revised code - but still using styles inline as they only apply to the Export page. Agree that ideally, they could be rolled in to the Dominion LESS files but perhaps we can assess the impact of doing that after QA? If we were to add to Dominion, then might have implications for users with custom themes?

jraddaoui replied:
Custom themes should extend Dominion and get this changes by default. There is also the possibility of a custom theme overwriting this template so it's better to do this changes in the existing LESS files from Dominion. And, in general, it's better to have separation of concerns.

PeteVox replied:
Have updated this. Added to Dominion scaffolding.

jraddaoui commented on apps/qubit/modules/settings/actions/clipboardAction.class.php:
```diff
@@ -45,7 +46,8 @@ public function earlyExecute()
       'clipboard_send_message_html' => $this->i18n->__('Sending...'),
-      'clipboard_send_http_method' => 'POST'
+      'clipboard_send_http_method' => 'POST',
+      'clipboard_export_digitalobjects_enabled' => '1',
```
I'd disable it by default but you may want to check with an analyst.

PeteVox replied:
Agree. Have changed to disabled by default, but will review with analyst.

jraddaoui commented on lib/job/arActorCsvExportJob.class.php:
```diff
@@ -35,6 +35,9 @@ class arActorCsvExportJob extends arBaseJob
   public function runJob($parameters)
   {
+    // Look for any digital object ids to export and initialize this->digitalObjects
+    $this->digitalObjects = ( array_key_exists('digitalObjects', $parameters) && !empty($parameters['digitalObjects'] ) ? $parameters['digitalObjects'] : array());
```
Cosmetic: avoid spaces inside and directly after/before the parenthesis.

jraddaoui commented on lib/job/arBaseJob.class.php:
```diff
@@ -284,34 +284,94 @@ private function createJobsDownloadsDirectory()
+                }
+                else
+                {
+                  // filename has been used - increment counter and add brackets
+                  $fileNames[$fileName]++;
+                  $doPathInfo = pathinfo( $doPath );
+                  $fileName = $doPathInfo['filename'] . ' (' . $fileNames[$fileName] . ').' . $doPathInfo['extension'];
```
Nicely thought! I'd prefer `filename_n.ext` instead of spaces and parenthesis.

jraddaoui commented on lib/job/arBaseJob.class.php:
```diff
@@ -284,34 +284,94 @@ private function createJobsDownloadsDirectory()
+                }
+                catch (Exception $e)
+                {
+                  if ($this->user->isAdministrator())
+                  {
+                    $errors[] = ('Exception: '.$e->getMessage());
+                  }
```
Why the administrator check here and not in the previous two possible errors?

PeteVox replied:
Have expanded this in revised code. Administrators should see full Exception (message), but any other user will see some more helpful text. To be reviewed.

jraddaoui reviewed (changes requested):
This is looking really nice @PeteVox, great work!

I have added quite a few comments bellow, most of them cosmetic and expected in a first contribution. We're working on improving this with a syntax checker/fixer in the near future (https://projects.artefactual.com/issues/13317).

There are a couple of important points about the DO processing in the request cycle and the JS code that interacts with the export form, but I'll add this feedback now and address those later in a different comment.

Thanks!!

jraddaoui commented:
Hi @PeteVox,

Sorry for the delay in the follow-up, the other two points I wanted to consider:

**DO checks in the request cycle:**

It won't be the most usual case, but we mentioned the other day the case of having a big amount of resources in the clipboard and having to check the existence of DOs and their permissions in the request cycle only to determine if the check-box is needed. My suggestion would be move those checks to the background process and always show the check-box, we do the same for the other check-boxes and this way you'll only need to move the code to the job. Please, check if that's possible with the assigned analyst and if it's not, I'd suggest that you break the loop as soon as you find an available DO, to avoid looping over all resources. This could get tricky if you want to avoid duplicated code and that's one of the reasons of why I prefer to just not check in the request cycle.

**Default `objectType` and involved Javascript:**

I've noticed that the "object/export" URL is open and that you can access it without and `objectType` parameter. I'm not even sure if that was working properly before, AFAIK you can access the URL but there is no actual link in AtoM without the parameter. Nevertheless, it's not working properly with the new behavior and I understand that you are going to make changes in there to make some check-boxes enabled/disabled based in their value. I'd appreciate if you can take a look at this issue, it may be as easy as adding a default value when the parameter is obtained to avoid the multiple options in the `objectType` field:

https://github.com/artefactual/atom/blob/qa/2.6.x/apps/qubit/modules/object/actions/exportAction.class.php#L39

```php
$this->objectType = $request->getParameter('objectType', 'informationObject');
```

Bests,
Radda.

jraddaoui commented on apps/qubit/modules/default/actions/alertsComponent.class.php:
```diff
@@ -43,22 +43,34 @@ public function execute($request)
+          }
+          else
+          {
+            $message .= ' ' . $this->context->i18n->__('%1%Refresh the page%2% for progress updates.', 
```
Nice catch!

jraddaoui commented on apps/qubit/modules/object/templates/exportSuccess.php:
```diff
@@ -17,63 +17,105 @@
+                  echo $form->includeAllLevels->renderRow();
+                }
+
+              ?>
```
Cosmetic: use the same templates format ...
```
<?php if (...): ?>
  <?php echo ...?>
<?php endif; ?>
```

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -69,54 +157,142 @@ protected function addField($name)
+          $this->helpMessage['30'] = __('Choosing "Include draft records" will include those marked with a Draft publication status in the export. Note: if you do NOT choose this option, any descendants of a draft record will also be excluded, even if they are published.');
+          $this->form->setDefault('includeDrafts', true);
+        }
+        else
+        {
+          $this->form->setWidget('includeDrafts', new sfWidgetFormInputHidden(array('label' => false)));
+        }
```
Are this hidden fields really needed? Could we just use `isset()` in the templates to check the field existence?

PeteVox replied:
Could do that Radda - but was trying to avoid conditional logic in the templates where possible. No problem to remove the hiddens and go with if/then in template if preferred?

PeteVox replied:
Didn't fully realise that isset() is ok to use here - thought that non-defined fields might impact any validators. Have updated in form and action class. Much cleaner.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -23,33 +23,121 @@ class ObjectExportAction extends DefaultEditAction
+
     $this->title = $this->context->i18n->__('Export');
 
+    if (0 != count($this->helpMessage))
+    {
+      // Convert sorted help message array to string
+      ksort($this->helpMessage);
+      $this->helpMessage = '<p>' . implode('</p><p>', $this->helpMessage) . '</p>';
+    }
```
What's the reason for this help to this help messages manipulation? I think it's clearer to have them associated to a field as they were or just add them in the template.

PeteVox replied:
Did initially add them to the relevant field (properties) in the first version, but Analyst team preferred the single pop-up help panel - for most fields. We need to build the help message conditionally depending on view/permissions/settings and I usually stack this type of thing in an array, rather than concatenating a string. No problem to change this to use string if preferred.

jraddaoui replied:
I think you could render the help messages in a different container, but I see your point. I'd prefer to build the array without keys (you can modify the order of the $names array to get the messages order needed) and the use a foreach in the template appending the paragraph tags, that way there's no need to access sf_data nor unescaping, and I think it's a bit more clear in the code.

PeteVox replied:
Have updated in repo. Removed keys and used foreach in template instead of constructing in controller. Had to still use `$sf_data->getRaw` as the array is `sfOutputEscaperArrayDecorator` object in template. Can update this if there's a better way to iterate.

jraddaoui replied:
Awesome, thanks @PeteVox!

jraddaoui commented on apps/qubit/modules/object/templates/exportSuccess.php:
```diff
@@ -17,63 +17,105 @@
+                  echo $form->includeDrafts->renderRow();
+                }
+
+              ?>
```
Same here.

jraddaoui commented on apps/qubit/modules/object/templates/exportSuccess.php:
```diff
@@ -17,63 +17,105 @@
+                }
+
+              ?>
+              <?php if (!empty($helpMessage)): ?>
+              <div class="alert alert-info generic-help animateNicely">
+                <?php echo(sfOutputEscaper::unescape($helpMessage)); ?>
+              </div>
               <?php endif; ?>
```
If you refactor the messages manipulation, even if there shouldn't be an issue considering its content, it's better to avoid the unescaping.

PeteVox replied:
Absolutely agree about the unescaping - but had to do something as the help message contains `<p></p>` tags. Did some more digging and found an alternative method using `$sf_data`.

jraddaoui replied:
A lot nicer, thanks! But see my comment above.

jraddaoui commented on js/exportOptions.js:
```diff
@@ -4,26 +4,32 @@
+      this.$formatSelect = this.$element.find('select[name="format"]');
+      this.formatDescriptionCsv = this.$formatSelect.find('option[value="csv"]').text();
+      this.formatDescriptionXml = this.$formatSelect.find('option[value="xml"]').text();
+      this.$includeAllLevels = this.$element.find('input[name="includeAllLevels"]:not([type="hidden"])');
```
Similar in here, you could rely on the existence instead of the type.

jraddaoui commented on js/exportOptions.js:
```diff
@@ -33,46 +39,119 @@
-      this.$includeDescendants.on('change', $.proxy(this.onIncludeDescendantsChange, this));
-      this.$includeAllLevels.on('change', $.proxy(this.onIncludeAllLevelsChange, this));
-      this.$objectType.on('change', $.proxy(this.onObjectTypeChange, this));
+      this.$type.on('change', $.proxy(this.onObjectTypeChange, this));
```
I think this is no longer possible with the new changes. Otherwise, we may be duplicating the logic between the controller and the Javascript.

PeteVox replied:
Think we're ok here. The `.js` doesn't change visibility of fields in the options panel any longer - this all done by the controller on page refresh depending on permissions, but also on whether there are valid objects of the selected type on the clipboard that can be exported. This logic also fixes attempts to directly access (and manipulate parameters of) the export page.

jraddaoui replied:
Sounds good, thanks @PeteVox. This is likely going to change after moving the clipboard content to the browser local storage, so it's more than okay as it is.

jraddaoui commented on js/exportOptions.js:
```diff
@@ -33,46 +39,119 @@
+      {
+        this.setDescendantsState(true);
+        this.setDigitalObjectsState(false);
+        this.$includeAllLevelsHolder.slideDown(this.animationMS);
+      }
+      else
+      {
+        var o = this;
+        this.$includeAllLevelsHolder.slideUp(this.animationMS).promise().done(function()
+        {
+          o.resetLevelsOptions();
+        });
+        this.setDigitalObjectsState(true);
+      }
     },
 
-    onIncludeDescendantsChange: function ()
+    onIncludeDigitalObjectsChange: function ()
     {
-      if (this.$includeDescendants.prop('checked'))
+      if (this.$includeDigitalObjects.prop('checked'))
       {
-        this.$includeAllLevels.parent().show();
+        if (!this.$includeAllLevelsHolder.is(":animated"))
+        {
+          var o = this;
+          this.$includeAllLevelsHolder.slideUp(this.animationMS).promise().done(function()
+          {
+            o.resetLevelsOptions();
+          });
+        }
+        this.setDescendantsState(false);
+        this.setDigitalObjectsState(true);
       }
       else
       {
-        this.resetLevelsOptions();
+        this.setDescendantsState(true);
+      }
+    },
```
Cosmetic: this two listeners could have been implemented in one, adding an extra argument to the function to know the direction of the switch.

jraddaoui replied:
@PeteVox, if you're not on this, please ignore. As mentioned, this is likely going to change after moving the clipboard content to the browser local storage.

PeteVox replied:
Have combined these in dev code.

jraddaoui commented on js/exportOptions.js:
```diff
@@ -33,46 +39,119 @@
+      }
+    },
+
+    setDescendantsState: function(b)
+    {
+      var o = this.$includeDescendants.parent();
+      if (b)
+      {
+        o.removeClass('muted');
+      }
+      else
+      {
+        this.$includeDescendants.attr('checked', false);
+        o.addClass('muted');
+      }
+    },
+
+    setDigitalObjectsState: function(b)
+    {
+      if (0 != this.$includeDigitalObjects.length)
+      {
+        var o = this.$includeDigitalObjects.parent();
+        if (b)
+        {
+          o.removeClass('muted');
+        }
+        else
+        {
+          this.$includeDigitalObjects.attr('checked', false);
+          o.addClass('muted');
+        }
       }
     },
```
Cosmetic: similar in here.

jraddaoui replied:
Same, ignore this part.

PeteVox replied:
Would prefer to leave as they are for the moment if that's ok - until we've been able to fully discuss the UI with the Analyst team. Once we know for sure if we need to make any significant changes to the logic that's applied to the visibility of the digital objects check box, can streamline the functions.

jraddaoui commented on js/exportOptions.js:
```diff
@@ -81,71 +160,28 @@
-      this.$levelDiv.addClass('hidden');
-      this.$includeAllLevels.attr('checked', true);
-      this.$levelSelect.val('');
+      window.location.href = url;
```
Is this switch still possible?

PeteVox replied:
This just refreshes the export page, but passes type so that controller can rebuild the options panel depending on permissions etc.

jraddaoui commented on lib/job/arBaseJob.class.php:
```diff
@@ -284,34 +284,120 @@ private function createJobsDownloadsDirectory()
    * Create ZIP file from results
    *
    * @param string  Path of file to write CSV data to
+   * @param string  Optional array of digital object ids to include
```
Cosmetic: no longer an array.

jraddaoui commented on lib/job/arBaseJob.class.php:
```diff
@@ -284,34 +284,120 @@ private function createJobsDownloadsDirectory()
+          // Get permitted digital object ids (if any) and iterate
+          foreach($this->getDigitalObjects() as $id)
+          {
+            if(!$error)
```
Maybe stop the loop entirely:

```
if ($error)
{
  break;
}

// Otherwise continue
```

jraddaoui commented on lib/job/arBaseJob.class.php:
```diff
@@ -284,34 +284,120 @@ private function createJobsDownloadsDirectory()
-      {
-        if (!is_dir($file))
+          // Get permitted digital object ids (if any) and iterate
+          foreach($this->getDigitalObjects() as $id)
```
Cosmetic: missing whitespace after `foreach`.

jraddaoui commented on lib/job/arBaseJob.class.php:
```diff
@@ -284,34 +284,120 @@ private function createJobsDownloadsDirectory()
+                          '%1%' => $id
+                        )
+                      );
+                      $error = true;
```
Should this be moved outside the `else` so it also acts on errors for admins?

PeteVox replied:
Have actually removed the error flags and using breaks instead as per suggestion. MUCH cleaner.

jraddaoui commented on lib/job/arBaseJob.class.php:
```diff
@@ -401,4 +487,87 @@ protected function canBeFullyExecuted()
+        case 'informationObject':
+          $criteria->addJoin(QubitInformationObject::ID, QubitSlug::OBJECT_ID);
+          // Hide drafts if necessary
+          if(!$this->params['public'])
```
Should this be the other way around? If it's public, then filter drafts?

PeteVox replied:
Nice catch. Fixed.

jraddaoui commented on lib/job/arBaseJob.class.php:
```diff
@@ -401,4 +487,87 @@ protected function canBeFullyExecuted()
+          // ----------
+          // Do appropriate ACL check(s). Master copy of text objects are always allowed for reading
+          // QubitActor does not have a ACL check for readmaster - so only enable for authenticated users.
+          if (
+            $digitalObject->masterAccessibleViaUrl()
+            && (
+              QubitTerm::TEXT_ID == $digitalObject->mediaTypeId
+              || (
+                'actor' == $this->params['objectType']
+                && $this->user->isAuthenticated()
+                && QubitAcl::check($item, 'read')
+              )
+              || (
+                'informationObject' == $this->params['objectType']
+                && QubitAcl::check($item, 'readMaster')
+                && QubitGrantedRight::checkPremis($item->id, 'readMaster')
+                && !$digitalObject->hasConditionalCopyright($item->id)
+              )
+            )
+          )
```
Really nice, quite and `if`!!

jraddaoui commented on lib/job/arRepositoryCsvExportJob.class.php:
```diff
@@ -55,7 +55,13 @@ public function runJob($parameters)
+    if (!empty($errors))
+    {
+      $this->error($this->i18n->__('Failed to create ZIP file.') . ' : ' . implode(' : ', $errors));
+      return false;
```
Cosmetic: empty line before return.

jraddaoui commented on lib/job/arInformationObjectXmlExportJob.class.php:
```diff
@@ -38,6 +38,9 @@ public function runJob($parameters)
     $this->params = $parameters;
 
+    // Do we include digital objects?
+    $this->includeDigitalObjects = array_key_exists('includeDigitalObjects', $parameters) && $parameters['includeDigitalObjects'];
```
Cosmetic: could we move this to `arBaseJob` using `$this->params` inside `createZipForDownload`? That way it could be removed from all export jobs and not passed to that function.

PeteVox replied:
Absolutely. Fixed. Much cleaner.

jraddaoui reviewed (changes requested):
Thanks @PeteVox!

This is looking great and it's pretty close to merge. I've added a few comments below, please let me know your thoughts about them before spending a lot of time as most of them are not really important.

My biggest concerns are:

- Moving the styles to the LESS files.
- Reduce form logic (hidden fields, duplicated logic in JS/controller/templates) and avoid messages array if possible.
- Drafts condition on export query.
- All errors for admins and only one for public users?

PeteVox committed:
Add DO option to clipboard export, refs #13395

Add a new system setting to enable export of digital objects attached to
the clipboard (information object and actor records only). All digital
object types can be exported, but only masters are exported for any
image type.

Digital objects are included in the export ZIP along with associated
metadata.

Incorporate basic user permissions checks prior to export, with further
checks for rights as part of the background job/task. Add a new method
to the digital object model to check for any conditional copyright
attached. If set, then do not export the affected digital object.

Modify the export page interface to rebuild options panel depending on
controller-based permissions checks. Also move help and other message
construction from template to controller.

Make the selection of descendant or digital objects in the export page
mutually exclusive. Add several new styling rules (LESS) to the Dominium
theme in scaffolding that apply to the options panel (#export-options).

Change some help and alert messages to improve consistency.

Build additional export form parameter validation in the controller to
mitigate attempts to manually override parameters, and also to prevent
direct access to the export page URL without objects on the clipboard.

jraddaoui reviewed (approved):
Thanks @PeteVox!

I've copied the final commit message to the PR title and description. Commit body is not wrapped at 72 chars but not a big deal, I'll merge.
